bin/ocamlformat-rpc/main.ml

```diff
@@ -58,63 +106,36 @@ let rec rpc_main = function
       | `Halt -> Ok ()
       | `Unknown | `Error _ -> rpc_main state
       | `Format x ->
-          List.fold_until ~init:()
-            ~finish:(fun () ->
-              V1.Command.output stdout
-                (`Error (Format.flush_str_formatter ())) )
-            ~f:(fun () try_formatting ->
-              match try_formatting conf x with
-              | Ok formatted ->
-                  ignore (Format.flush_str_formatter ()) ;
-                  V1.Command.output stdout (`Format formatted) ;
-                  Stop ()
-              | Error e ->
-                  Translation_unit.Error.print Format.str_formatter e ;
-                  Continue () )
-            (* The formatting functions are ordered in such a way that the
-               ones expecting a keyword first (like signatures) are placed
-               before the more general ones (like toplevel phrases). Parsing
-               a file as `--impl` with `ocamlformat` processes it as a use
-               file (toplevel phrases) anyway.
-
-               `ocaml-lsp` should use core types, module types and
-               signatures. `ocaml-mdx` should use toplevel phrases,
-               expressions and signatures. *)
-            [ format Core_type
-            ; format Signature
-            ; format Module_type
-            ; format Expression
-            ; format Use_file ] ;
+          ( match run_format conf x with
+          | `Format formatted ->
+              ignore (Format.flush_str_formatter ()) ;
+              V1.Command.output stdout (`Format formatted)
+          | `Error e -> handle_format_error e ) ;
           rpc_main state
+      | `Combined (file, c, x) ->
```

Comment:
This isn't much more extensible than the previous format command. What are your thoughts on:

- Add a "config_for_path" and "override_config" commands, that sets states within the server, the sequence of commands would be `("config_for_path", "path")` then `("format", "buffer")`.

- Take extra arguments to the format command that is a "sexp record" so it's extensible:
  ```sexp
  ( "format" ("config_for_path" "path") ("override_config" (config ...)) )
  ```

Reply:
I think that @rgrinberg whished to be able to configure and format in a single command for vs-code interaction with ocamlformat_rpc. So the goal of the Combined command wasn't really for extensibility. However, I agree that it would make sense to add a `Config_for_path` command. So that the `Combined` command is a combination of three commands: `Config_for_path`, `Config` and `Format` (instead of only the two latter + config for path but not available as a command).

I also see the advantage of being able to override the config, without modifying it permanently, in the format command. However, I'm not sure how much we can modify it, without bumping the protocol version...

Following your comment, I can add a `Config_for_path` command, and make the `Combined` command also takes an `overriding_config` that overrides but no modify the global config. I don't think that modifying the `Format` command is a good idea, as bumping the protocol version is a good idea, as I think it requires further thinking to design a better protocol.

What do you think?

Reply:
Modifying the "format" command can be done in a backward compatible way, I think.

It's currently matched as `List [Atom "Format"; Atom x]`: `(Format "path")`.
It is safe to add more arguments: `List (Atom "Format" :: Atom x :: args)`: `(Format "path" option1 option2)` where each options could be arbitrary sexps.

Reply:
We could also imagine a new command "Path" that could modify the local config similarly to Config. Vscode/ocamllsp would only call this command when the new file to reformat is in a different directory, it would be more efficient than recomputing the whole config every time a new file is loaded.

Reply:
The per-directory caching could pretty much be implemented in the server.

Reply:
Hi, I'd like a single command not because of performance (although I'll take always take the extra perf) but because I'd like to assume a little as possible about how ocamlformat works. I don't want to do any of:

* Load/Unload configuration from files
* Think about the lifetime of configuration values
* Be aware that configuration even exists

You might say I'm rejecting a more flexible API and limiting myself in the future. But the way I see it is that I'm reserving your project the maximum degrees of freedom to change the implementation without consulting me. Consequently, the lesser coupling of our projects grants our users more freedom to pick components without worry. That is valuable.